org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/refactoring2/ChangeConverter.java

```diff
@@ -0,0 +1,245 @@
+import org.eclipse.xtext.ide.serializer.IEmfResourceChange;
+import org.eclipse.xtext.ide.serializer.ITextDocumentChange;
+import org.eclipse.xtext.ui.refactoring.impl.EditorDocumentChange;
+import org.eclipse.xtext.ui.util.DisplayRunnableWithResult;
+import org.eclipse.xtext.util.Exceptions;
+import org.eclipse.xtext.util.IAcceptor;
+import org.eclipse.xtext.util.internal.Nullable;
+
+import com.google.common.base.Predicate;
+import com.google.inject.Inject;
+
+import static org.eclipse.xtext.ui.refactoring2.TryWithResource.tryWith;
+
+/**
+ * Converts {@link IEmfResourceChange}s to LTK {@link Change}s.
+ * 
+ * @author koehnlein - Initial contribution and API
+ * @since 2.13
+ */
+public class ChangeConverter implements IAcceptor<IEmfResourceChange> {
+
+	public static class Factory {
+
+		@Inject
+		private ResourceURIConverter resourceURIConverter;
+
+		@Inject(optional = true)
+		@Nullable
+		private IWorkbench workbench;
+
+		public ChangeConverter create(String name, Predicate<Change> changeFilter, RefactoringIssueAcceptor issues) {
+			return new ChangeConverter(name, changeFilter, issues, resourceURIConverter, workbench);
+		}
+	}
+
+	private final CompositeChange currentChange;
+
+	private final RefactoringIssueAcceptor issues;
+
+	private final Predicate<Change> changeFilter;
+
+	private final ResourceURIConverter resourceUriConverter;
+
+	private final IWorkbench workbench;
+
+	protected ChangeConverter(String name, Predicate<Change> changeFilter, RefactoringIssueAcceptor issues,
+			ResourceURIConverter uriConverter, IWorkbench workbench) {
+		currentChange = new CompositeChange(name);
+		this.issues = issues;
+		this.changeFilter = changeFilter;
+		this.resourceUriConverter = uriConverter;
+		this.workbench = workbench;
+	}
+
+	@Override
+	public void accept(IEmfResourceChange emfResourceChange) {
+		doConvert(emfResourceChange);
+	}
+
+	public Change getChange() {
+		if (currentChange.getChildren().length == 0) {
+			return null;
+		} else {
+			return currentChange;
+		}
+	}
+
+	protected void doConvert(IEmfResourceChange change) {
+		handleReplacements(change);
+		handleUriChange(change);
+		if (affectsPersistedFiles()) {
+			saveEditorsAfterApply();
+		}
+	}
+
+	protected void handleReplacements(IEmfResourceChange change) {
+		if (change instanceof ITextDocumentChange) {
+			_handleReplacements((ITextDocumentChange) change);
+		} else if (change != null) {
+			_handleReplacements(change);
+		} else {
+			throw new IllegalArgumentException("Unhandled parameter types: " + Arrays.asList(change).toString());
+		}
+	}
+
+	protected void _handleReplacements(IEmfResourceChange change) {
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		tryWith(outputStream, () -> {
+			try {
+				IFile file = resourceUriConverter.toFile(change.getOldURI());
+				if (!canWrite(file)) {
+					issues.add(RefactoringIssueAcceptor.Severity.ERROR, "Affected file '" + file.getFullPath() + "' is read-only");
```

Comment:
missing return? Even though the Xtend original simply ignores that diagnostic, too.

Reply:
Are you suggesting to **add** a return statement on line 128?

Reply:
Indeed. If the file is readonly, it’s probably bad to attempt a modification